test/tap/move-no-clobber-dest-node-modules.js

```diff
@@ -0,0 +1,135 @@
+var common = require('../common-tap.js')
+
+function fixturepath () {
+  return path.join(base, path.join.apply(path, arguments))
+}
+
+function File (contents) {
+  return {
+    type: 'file',
+    contents: typeof contents === 'object' ? JSON.stringify(contents) : contents,
+  }
+}
+
+function Dir (contents) {
+  return {
+    type: 'dir',
+    contents: contents || {}
+  }
+}
+
+function createFixtures (filename, fixture) {
+  if (fixture.type === 'dir') {
+    mkdirp.sync(filename)
+    Object.keys(fixture.contents).forEach(function (content) {
+      createFixtures(path.resolve(filename, content), fixture.contents[content])
+    })
+  } else if (fixture.type === 'file') {
+    fs.writeFileSync(filename, fixture.contents)
+  } else {
+    throw new Error("Unknown fixture type: " + fixture.type)
+  }
+}
+
+var fixtures = Dir({
+// The fixture modules
+  'moda@1.0.1': Dir({
+    'package.json': File({
+      name: 'moda',
+      version: '1.0.1'
+    })
+  }),
+  'modb@1.0.0': Dir({
+    'package.json': File({
+      name: 'modb',
+      version: '1.0.0',
+      bundleDependencies: ['modc'],
+      dependencies: {
+        modc: fixturepath('modc@1.0.0')
+      }
+    })
+  }),
+  'modc@1.0.0': Dir({
+    'package.json': File({
+      name: 'modc',
+      version: '1.0.0'
+    })
+  }),
+// The local config
+  'package.json': File({
+    dependencies: {
+      moda: fixturepath('moda@1.0.1'),
+      modb: fixturepath('modb@1.0.0')
+    }
+  }),
+  'node_modules': Dir({
+    'moda': Dir({
+      'package.json': File({
+        name: 'moda',
+        version: '1.0.0',
+        _requested: { rawSpec: fixturepath('moda@1.0.0') },
+        dependencies: {
+          modb: fixturepath("modb@1.0.0")
+        },
+        bundleDependencies: [ 'modb' ]
+      })
+    })
+  })
+})
+
+function setup () {
+  cleanup()
+  createFixtures(base, fixtures)
+}
+
+function cleanup () {
+  rimraf.sync(base)
+}
+
+function exists (t, filepath, msg) {
+  try {
+    fs.statSync(filepath)
+    t.pass(msg)
+    return true
+  } catch (ex) {
+    t.fail(msg, {found: null, wanted: 'exists', compare: 'fs.stat(' + filepath + ')'})
+    return false
+  }
+}
+
+test('setup', function (t) {
+  setup()
+  common.npm('install', {cwd: fixturepath('modb@1.0.0')}, function (err, code, stdout, stderr) {
+    if (err) throw err
+    t.is(code, 0, 'modb')
+    common.npm('install', {cwd: fixturepath('node_modules','moda')}, function (err, code, stdout, stderr) {
+      if (err) throw err
+      t.is(code, 0, 'installed moda')
+      t.done()
+    })
+  })
+})
+
+test('no-clobber', function (t) {
+  common.npm('install', {cwd: base}, function (err, code, stdout, stderr) {
+    if (err) throw err
+    t.is(code, 0, 'installed ok')
+    exists(t, fixturepath('node_modules','moda'), 'moda')
+    exists(t, fixturepath('node_modules', 'modb'), 'modb')
+    exists(t, fixturepath('node_modules', 'modb', 'node_modules', 'modc'), 'modc')
```

Comment:
Ok, so to start with after setup completes we have an existing node_modules that looks like this:
```
npm sill currentTree move-no-clobber-dest-node-modules
npm sill currentTree └─┬ moda@1.0.0
npm sill currentTree   └─┬ modb@1.0.0
npm sill currentTree     └── modc@1.0.0
```
The only caveat here is that `moda@1.0.0` bundles both `modb` and `modc`.

Now, our top level module has a `package.json` that depends on `moda@1.0.1` and `modb@1.0.0`. `moda@1.0.1` differs from `moda@1.0.0` in that it doesn't depend on anything.

So `npm install` needs to mutate that existing tree such that the result is valid according to the `package.json`.  With this patch that looks like:

```
npm sill idealTree move-no-clobber-dest-node-modules
npm sill idealTree ├── moda@1.0.1
npm sill idealTree └─┬ modb@1.0.0
npm sill idealTree   └── modc@1.0.0
```

The diff for this is:

```
npm sill diffTrees remove modc@1.0.0
npm sill diffTrees add modc@1.0.0
npm sill diffTrees remove modb@1.0.0
npm sill diffTrees update moda@1.0.1
npm sill diffTrees add modb@1.0.0
```

Without this patch, the diff is:

```
npm sill diffTrees remove modc@1.0.0
npm sill diffTrees add modc@1.0.0
npm sill diffTrees update moda@1.0.1
npm sill diffTrees move modb@1.0.0
```

That looks ok to me, so I'm a bit baffled too. So long as our order of operations is ok then that shoul dwork fine. So let's dig further into the differences between the decomposed actions. The following is from the patched version:

```
fetch moda@1.0.1, modb@1.0.0
extract moda@1.0.1, modb@1.0.0
preinstall modc@1.0.0, moda@1.0.1, modb@1.0.0
remove moda@1.0.1
remove modb@1.0.0
remove modc@1.0.0
finalize modc@1.0.0
finalize moda@1.0.1
finalize modb@1.0.0
build modc@1.0.0
build moda@1.0.1
build modb@1.0.0
install modc@1.0.0
install moda@1.0.1
install modb@1.0.0
postinstall modc@1.0.0
postinstall moda@1.0.1
postinstall modb@1.0.0
```

And this is from the unpatched version:

```
fetch moda@1.0.1
extract moda@1.0.1
preinstall modc@1.0.0, moda@1.0.1
remove moda@1.0.1
remove modc@1.0.0
move modb@1.0.0
finalize modc@1.0.0
finalize moda@1.0.1
build modc@1.0.0
build moda@1.0.1
build modb@1.0.0
install modc@1.0.0
install moda@1.0.1
install modb@1.0.0
postinstall modc@1.0.0
postinstall moda@1.0.1
postinstall modb@1.0.0
```

The unpatched version is for some reason skipping fetch/extract for `modc` which is why it doesn't exist when we go to `finalize` (we get an ENOENT on rename of a `.staging/modulename-deadbeef` folder).

Ok, so why _is_ that? Well, `decompose-actions` has this code:

```
  if (!pkg.fromBundle) {
    decomposed.push(['fetch', pkg])
    decomposed.push(['extract', pkg])
    decomposed.push(['test', pkg])
  }
```

And it exists because if we're installing a module FROM a bundle then it's included in that other module's tarball and thus shouldn't be fetched/extracted on its own. That makes sense, what's baffling here however is WHY `fromBundled` would be set in this case as that's an attribute that's ordinarily ONLY set when installing a new module with bundled dependencies, and none of those `add`'s include bundles.

Digging in here, I think the patch this is testing is just papering over a deeper problem and that may be why I sat on this last year and didn't push it out. But now we have a record of WHY at least. =D